File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/system/equipment/anesthesia/SEAnesthesiaMachineOxygenBottle.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

package mil.tatrc.physiology.datamodel.system.equipment.anesthesia;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.properties.SEScalarVolume;
import mil.tatrc.physiology.datamodel.bind.AnesthesiaMachineOxygenBottleData;

public class SEAnesthesiaMachineOxygenBottle
{
    protected SEScalarVolume volume;
    
    public SEAnesthesiaMachineOxygenBottle()
    {
        volume = null;
    }
    
    public void reset()
    {
        if (volume != null)
            volume.invalidate();
    }
    
    public void copy(SEAnesthesiaMachineOxygenBottle from)
    {
      reset();
      if(from.hasVolume())
        this.getVolume().set(from.getVolume());
    }
    
    public boolean load(AnesthesiaMachineOxygenBottleData in)
    {
        if (in.getVolume() != null)
            getVolume().load(in.getVolume());
        
        return true;
    }
    
    public AnesthesiaMachineOxygenBottleData unload()
    {
        AnesthesiaMachineOxygenBottleData data
            = CDMSerializer.objFactory.createAnesthesiaMachineOxygenBottleData();
        unload(data);
        
        return data;
    }
    
    protected void unload(AnesthesiaMachineOxygenBottleData data)
    {
        if (hasVolume())
            data.setVolume(volume.unload());
    }
    
    public boolean hasVolume()
    {
        return volume == null ? false : volume.isValid();
    }
    
    public SEScalarVolume getVolume()
    {
        if (volume == null)
            volume = new SEScalarVolume();        
        return volume;
    }
    
    public String toString()
  	{
  			return "Anesthesia Machine Oxygen Bottle"
  			+ "\n\tVolume: " + getVolume();
  	}
}

File: projects/biogears/libBiogears/src/cdm/system/equipment/Anesthesia/SEAnesthesiaMachineOxygenBottle.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include <biogears/cdm/system/equipment/Anesthesia/SEAnesthesiaMachineOxygenBottle.h>

#include "io/cdm/Anesthesia.h"
#include "io/cdm/AnesthesiaActions.h"

#include <biogears/cdm/properties/SEScalarVolume.h>
#include <biogears/cdm/substance/SESubstanceManager.h>
#include <biogears/container/Tree.tci.h>

namespace biogears {
namespace io {
  class Anesthesia;
}
SEAnesthesiaMachineOxygenBottle::SEAnesthesiaMachineOxygenBottle(Logger* logger)
  : Loggable(logger)
{
  m_Volume = nullptr;
}
//-------------------------------------------------------------------------------

SEAnesthesiaMachineOxygenBottle::~SEAnesthesiaMachineOxygenBottle()
{
  Clear();
}
//-------------------------------------------------------------------------------

void SEAnesthesiaMachineOxygenBottle::Clear()
{
  SAFE_DELETE(m_Volume);
}
//-------------------------------------------------------------------------------

bool SEAnesthesiaMachineOxygenBottle::Load(const CDM::AnesthesiaMachineOxygenBottleData& in)
{
  io::Anesthesia::UnMarshall(in, *this);
  return true;
}
//-------------------------------------------------------------------------------

CDM::AnesthesiaMachineOxygenBottleData* SEAnesthesiaMachineOxygenBottle::Unload() const
{
  CDM::AnesthesiaMachineOxygenBottleData* data = new CDM::AnesthesiaMachineOxygenBottleData();
  Unload(*data);
  return data;
}
//-------------------------------------------------------------------------------

void SEAnesthesiaMachineOxygenBottle::Unload(CDM::AnesthesiaMachineOxygenBottleData& data) const
{
  io::Anesthesia::Marshall(*this, data);
}
//-------------------------------------------------------------------------------

void SEAnesthesiaMachineOxygenBottle::Merge(const SEAnesthesiaMachineOxygenBottle& from)
{
  COPY_PROPERTY(Volume);
}
//-------------------------------------------------------------------------------
const SEScalar* SEAnesthesiaMachineOxygenBottle::GetScalar(const char* name)
{
  return GetScalar(std::string { name });
}
//-------------------------------------------------------------------------------
const SEScalar* SEAnesthesiaMachineOxygenBottle::GetScalar(const std::string& name)
{
  if (name == "Volume")
    return &GetVolume();
  return nullptr;
}
//-------------------------------------------------------------------------------

bool SEAnesthesiaMachineOxygenBottle::HasVolume() const
{
  return m_Volume == nullptr ? false : m_Volume->IsValid();
}
//-------------------------------------------------------------------------------

SEScalarVolume& SEAnesthesiaMachineOxygenBottle::GetVolume()
{
  if (m_Volume == nullptr)
    m_Volume = new SEScalarVolume();
  return *m_Volume;
}
//-------------------------------------------------------------------------------
double SEAnesthesiaMachineOxygenBottle::GetVolume(const VolumeUnit& unit) const
{
  if (m_Volume == nullptr)
    return SEScalar::dNaN();
  return m_Volume->GetValue(unit);
}
//-------------------------------------------------------------------------------
void SEAnesthesiaMachineOxygenBottle::ToString(std::ostream& str)
{
  str << " Anesthesia Machine Oxygen Bottle, ";
  if (HasVolume())
    str << "\rVolume : " << GetVolume();
  else
    str << "\rVolume : NaN" << GetVolume();
  str << std::flush;
}
//-------------------------------------------------------------------------------
bool SEAnesthesiaMachineOxygenBottle::operator==(SEAnesthesiaMachineOxygenBottle const& rhs) const
{
  return (m_Volume && rhs.m_Volume) ? m_Volume->operator==(*rhs.m_Volume) : m_Volume == rhs.m_Volume;
}
//-------------------------------------------------------------------------------
bool SEAnesthesiaMachineOxygenBottle::operator!=(SEAnesthesiaMachineOxygenBottle const& rhs) const
{
  return !(*this == rhs);
}
//-------------------------------------------------------------------------------
}